sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java

```diff
@@ -889,28 +886,33 @@ public void processElement(ProcessContext c) {
         batchCount++;
 
         if (batchCount >= DEFAULT_BATCH_SIZE) {
-          pipeline.exec();
-          pipeline.sync();
-          pipeline.multi();
+          transaction.exec();
+          transaction.multi();
           batchCount = 0;
         }
       }
 
       private void writeRecord(KV<String, Map<String, String>> record) {
         String key = record.getKey();
         Map<String, String> value = record.getValue();
+        final XAddParams commonParams = new XAddParams().id(StreamEntryID.NEW_ENTRY);
+        final XAddParams params;
         if (spec.maxLen() > 0L) {
-          pipeline.xadd(key, StreamEntryID.NEW_ENTRY, value, spec.maxLen(), spec.approximateTrim());
+          if (spec.approximateTrim()) {
+            params = commonParams.maxLen(spec.maxLen()).approximateTrimming();
+          } else {
+            params = commonParams.maxLen(spec.maxLen());
+          }
         } else {
-          pipeline.xadd(key, StreamEntryID.NEW_ENTRY, value);
+          params = commonParams;
```

Comment:
I'd suggest to simplify this. it's a mutable fluent api, so you could just write
```
        final XAddParams params = new XAddParams().id(StreamEntryID.NEW_ENTRY);
        if (spec.maxLen() > 0L) {
          params.maxLen(spec.maxLen());
          if (spec.approximateTrim()) {
            params.approximateTrimming();
          }
        }
```